src/communication/NetworkClient.ts

```diff
@@ -161,28 +145,76 @@ export default class NetworkClient {
-      // inaccurate guesses do not result in wastefully short pages.)
-      let limit: number;
+      // Pick the page sizes (limits) based on the guessed demand.
+      let popLimit: () => number;
+      // If no demand was guessed, pages of 128 values each will be requested.
       if (demand == Number.POSITIVE_INFINITY) {
-        limit = 128;
+        popLimit = () => 128;
       } /* if (demand != Number.POSITIVE_INFINITY) */ else {
-        limit = Math.max(Math.ceil(demand / Math.ceil(demand / 250)), 64);
+        let pageSizes: number[];
+        // If a demand of 64 or less was guessed, request an initial page of 64 values. This ensures inaccurate guesses
+        // do not result in wastefully short pages.
+        if (demand <= 64) {
+          pageSizes = [64];
+        } /* if (demand > 64 && demand != Number.POSITIVE_INFINITY) */ else {
+          // If a demand of over 64 was guessed, request pages of 250 values ‒ a limit imposed by the Mollie API ‒
+          // until the remaining demand can be met with a single page. Finally, request that single page.
+          pageSizes = new Array(Math.ceil(demand / 250));
+          pageSizes[0] = demand - (pageSizes.length - 1) * 250;
+          pageSizes.fill(250, 1);
+        }
+        // If more values are requested after the guessed demand (the guess was inaccurate), request pages of 128
+        // values each.
+        popLimit = () => pageSizes.pop() ?? 128;
       }
-      return new HelpfulIterator<R>(iterate(this.axiosInstance, type, `${relativePath}${stringifyQuery({ ...query, limit })}`));
+      // Set up the throttler. The iterator will wait before making a request to the Mollie API if more values have
+      // been consumed than "allowed" by the valuesPerMinute argument. Note that this iterator will not interrupt the
+      // yielding of values which have already been received from the Mollie API. If a page of 250 values is received
+      // and valuesPerMinute is set to 100, all 250 values received values will be yielded before the (two-minute)
+      // break.
+      const throttler = new Throttler(valuesPerMinute);
+      const { axiosInstance } = this;
+      return new HelpfulIterator<R>(
+        (async function* iterate<R>() {
+          let url = buildUrl(pathname, { ...query, limit: popLimit() });
+          while (true) {
+            // Request and parse the page from the Mollie API.
+            const response = await axiosInstance.get(url).catch(throwApiError);
+            try {
+              /* eslint-disable-next-line no-var */
+              var { _embedded: embedded, _links: links } = response.data;
+            } catch (error) {
+              throw new ApiError('Received unexpected response from the server');
+            }
+            // Yield the values on the page.
+            const values = embedded[binderName] as R[];
+            yield* values;
```

Comment:
This is more of a question for my understanding, because I probably have a thinking error somewhere:
Shouldn't `throttler.tally` be called before `yield`?

From what I understand the generator function will be halted at `yield` on every call of `.next()`.
So currently it would halt before tallying the values. It would only tally on the subsequent `next()`  call (after depleting values). However if enough time has passed between the previous call and this one already, the values wouldn't have to be throttled anymore, right? 🤔

Reply:
I'm not 100% sure whether I'm following your reasoning, but let's see.

`tally` doesn't throttle. [It keeps track of consumption](https://github.com/mollie/mollie-api-node/pull/283/files#diff-97360564710ef1f5c97fb64333a27c7bbde04935a821651458a7d8877f215b3aR22). The `throttle` function throttles if there is any overconsumption. So while it is true that putting `tally` before or after `yield` can make a big difference in when it was called, the effect is the same.

I think I should maybe document this better, but the idea is _not_ that the throttler throttles the yielding. If 100 payments were downloaded, but `valuesPerMinute` is set to 50, it just yields all of them immediately… but waits a bit before making the _next_ request. Only yielding 50 values even though 100 are available seems a bit childish and unproductive.

Does any of this clear anything up?